flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java

```diff
@@ -118,27 +138,168 @@ public static JobExecutionResult submitJobAndWait(
+		Props jobClientActorProps = JobClientActor.createJobClientActorProps(
+			leaderRetrievalService,
+			timeout,
+			sysoutLogUpdates);
+
+		ActorRef jobClientActor = actorSystem.actorOf(jobClientActorProps);
+
+		Future<Object> attachmentFuture = Patterns.ask(
+				jobClientActor,
+				new JobClientMessages.AttachToJobAndWait(jobID),
+				new Timeout(AkkaUtils.INF_TIMEOUT()));
+
+		return new JobListeningContext(
+				jobID,
+				attachmentFuture,
+				jobClientActor,
+				classloader);
+	}
+
+	/**
+	 * Reconstructs the class loader by first requesting information about it at the JobManager
+	 * and then downloading missing jar files.
+	 * @param jobID id of job
+	 * @param jobManager gateway to the JobManager
+	 * @param config the flink configuration
+	 * @param timeout timeout for querying the jobmanager
+	 * @return A classloader that should behave like the original classloader
+	 * @throws JobRetrievalException if anything goes wrong
+	 */
+	public static ClassLoader retrieveClassLoader(
+		JobID jobID,
+		ActorGateway jobManager,
+		Configuration config,
+		FiniteDuration timeout)
+		throws JobRetrievalException {
+
+		BlobCache blobClient = null;
+		try {
+			final Object jmAnswer;
+			try {
+				jmAnswer = Await.result(
+					jobManager.ask(
+						new JobManagerMessages.RequestClassloadingProps(jobID), timeout), timeout);
+			} catch (Exception e) {
+				throw new JobRetrievalException(jobID, "JobManager didn't respond", e);
+			}
+
+			if (jmAnswer instanceof JobManagerMessages.ClassloadingProps) {
+				JobManagerMessages.ClassloadingProps props = ((JobManagerMessages.ClassloadingProps) jmAnswer);
+
+				Option<String> jmHost = jobManager.actor().path().address().host();
+				String jmHostname = jmHost.isDefined() ? jmHost.get() : "localhost";
+				InetSocketAddress serverAddress = new InetSocketAddress(jmHostname, props.blobManagerPort());
+				blobClient = new BlobCache(serverAddress, config);
+
+				final List<BlobKey> requiredJarFiles = props.requiredJarFiles();
+				final List<URL> requiredClasspaths = props.requiredClasspaths();
+
+				final URL[] allURLs = new URL[requiredJarFiles.size() + requiredClasspaths.size()];
+
+				int pos = 0;
+				for (BlobKey blobKey : props.requiredJarFiles()) {
+					try {
+						allURLs[pos++] = blobClient.getURL(blobKey);
+					} catch (Exception e) {
+						throw new JobRetrievalException(jobID, "Failed to download BlobKey " + blobKey);
+					}
+				}
+
+				for (URL url : requiredClasspaths) {
+					allURLs[pos++] = url;
+				}
+
+				return new URLClassLoader(allURLs, JobClient.class.getClassLoader());
+			} else if (jmAnswer instanceof JobManagerMessages.JobNotFound) {
+				throw new JobRetrievalException(jobID, "Couldn't retrieve class loader. Job " + jobID + " not found");
+			} else {
+				throw new JobRetrievalException(jobID, "Unknown response from JobManager: " + jmAnswer);
+			}
+		} finally {
+			if (blobClient != null) {
+				blobClient.shutdown();
```

Comment:
Yes thank you! This should only be called if the downloading fails. Otherwise, we use the blobclient's shutdown hook.